### PR TITLE
[DebugInfo] Fix verifier crash for complex switch 

### LINKED ISSUE #73338: Building postgres-nio crashes 6.0 nightly
### Description

If I build https://github.com/vapor/postgres-nio with the latest 6.0 nightly the compiler asserts with a `SIL verification failed: Two variables with different type but same scope!` error

The compiler output is so large I can't add it into the issue so I have included a link to it
[postgres-nio-compiler-crash.txt](https://github.com/apple/swift/files/15165613/postgres-nio-compiler-crash.txt)

### Reproduction

```swift
git clone https://github.com/vapor/postgres-nio
cd postgres-nio
swift build
```

### Stack dump

```text
Compiler output is so large I have included the start and the end

error: compile command failed due to signal 6 (use -v to see invocation)
SIL verification failed: Two variables with different type but same scope!: lhs == rhs || (lhs.isAddress() && lhs.getObjectType() == rhs) || (DebugVarTy.isAddress() && lhs == rhs.getObjectType())
Verifying instruction:
->   %372 = alloc_stack [lexical] [var_decl] $TinyFastSequence<TimerCancellationToken> // users: %392, %391, %401, %400, %386, %373
     copy_addr [take] %371 to [init] %372 : $*TinyFastSequence<TimerCancellationToken> // id: %373
     %386 = apply %385<TinyFastSequence<TimerCancellationToken>>(%382, %372) : $@convention(thin) <τ_0_0 where τ_0_0 : Equatable> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> Bool // user: %389
     destroy_addr %372 : $*TinyFastSequence<TimerCancellationToken> // id: %391
     dealloc_stack %372 : $*TinyFastSequence<TimerCancellationToken> // id: %392
     destroy_addr %372 : $*TinyFastSequence<TimerCancellationToken> // id: %400
     dealloc_stack %372 : $*TinyFastSequence<TimerCancellationToken> // id: %401
In function:
// static PoolStateMachine.ConnectionAction<>.== infix(_:_:)
sil [ossa] @$s21_ConnectionPoolModule0B12StateMachineV0A6ActionOAASQR3_rlE2eeoiySbAEyxq_q0_q1_q2_q3__G_AGtFZ : $@convention(method) <Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken where Connection : PooledConnection, ConnectionIDGenerator : ConnectionIDGeneratorProtocol, ConnectionID == Connection.ID, Request : ConnectionRequestProtocol, RequestID == Request.ID, TimerCancellationToken : Equatable, TimerCancellationToken : Sendable, Connection.ID == ConnectionIDGenerator.ID> (@in_guaranteed PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, @in_guaranteed PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, @thin PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction.Type) -> Bool {
// %0 "lhs"                                       // users: %7, %3
// %1 "rhs"                                       // users: %9, %4
// %2 "self"                                      // user: %5
bb0(%0 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, %1 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, %2 : $@thin PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction.Type):
  debug_value %0 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, let, name "lhs", argno 1, expr op_deref // id: %3
  debug_value %1 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, let, name "rhs", argno 2, expr op_deref // id: %4
  debug_value %2 : $@thin PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction.Type, let, name "self", argno 3 // id: %5
  %6 = alloc_stack $PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction // users: %452, %433, %398, %367, %353, %325, %294, %264, %198, %132, %72, %44, %13, %7
  copy_addr %0 to [init] %6 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction // id: %7
  %8 = alloc_stack $PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction // users: %451, %432, %397, %366, %352, %324, %293, %263, %197, %131, %71, %43, %14, %9
  copy_addr %1 to [init] %8 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction // id: %9
  %10 = alloc_stack $(PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction) // users: %450, %449, %431, %396, %365, %351, %323, %292, %262, %196, %130, %70, %42, %16, %15, %12, %11
  %11 = tuple_element_addr %10 : $*(PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction), 0 // user: %13
  %12 = tuple_element_addr %10 : $*(PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction), 1 // user: %14
  copy_addr [take] %6 to [init] %11 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction // id: %13
  copy_addr [take] %8 to [init] %12 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction // id: %14
  %15 = tuple_element_addr %10 : $*(PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction), 0 // users: %430, %395, %364, %350, %322, %300, %291, %271, %261, %205, %195, %139, %129, %82, %69, %41, %19, %17
  %16 = tuple_element_addr %10 : $*(PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction), 1 // users: %429, %405, %394, %370, %363, %362, %349, %321, %304, %302, %275, %274, %211, %210, %145, %144, %87, %86, %68, %40, %23, %21
  switch_enum_addr %15 : $*PoolStateMachine<Connection, ConnectionIDGenerator, ConnectionID, Request, RequestID, TimerCancellationToken>.ConnectionAction, case #PoolStateMachine.ConnectionAction.scheduleTimers!enumelt: bb1, case #PoolStateMachine.ConnectionAction.makeConnection!enumelt: bb7, case #PoolStateMachine.ConnectionAction.runKeepAlive!enumelt: bb11, case #PoolStateMachine.ConnectionAction.closeConnection!enumelt: bb15, case #PoolStateMachine.ConnectionAction.shutdown!enumelt: bb19, case #PoolStateMachine.ConnectionAction.cancelTimers!enumelt: bb22, case #PoolStateMachine.ConnectionAction.none!enumelt: bb28 // id: %17

End of compiler output

Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.      Program arguments: /usr/bin/swift-frontend -frontend -c /workspaces/postgres-nio/Sources/ConnectionPoolModule/ConnectionIDGenerator.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/ConnectionPool.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/ConnectionPoolError.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/ConnectionPoolObservabilityDelegate.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/ConnectionRequest.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/Max2Sequence.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/NIOLock.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/NIOLockedValueBox.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/NoKeepAliveBehavior.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine+ConnectionState.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine+RequestQueue.swift -primary-file /workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine.swift /workspaces/postgres-nio/Sources/ConnectionPoolModule/TinyFastSequence.swift -emit-dependencies-path /workspaces/postgres-nio/.build/x86_64-unknown-linux-gnu/debug/_ConnectionPoolModule.build/PoolStateMachine.d -emit-reference-dependencies-path /workspaces/postgres-nio/.build/x86_64-unknown-linux-gnu/debug/_ConnectionPoolModule.build/PoolStateMachine.swiftdeps -target x86_64-unknown-linux-gnu -disable-objc-interop -I /workspaces/postgres-nio/.build/x86_64-unknown-linux-gnu/debug/Modules -color-diagnostics -enable-testing -g -debug-info-format=dwarf -dwarf-version=4 -module-cache-path /workspaces/postgres-nio/.build/x86_64-unknown-linux-gnu/debug/ModuleCache -swift-version 5 -Onone -D SWIFT_PACKAGE -D DEBUG -empty-abi-descriptor -resource-dir /usr/lib/swift -enable-anonymous-context-mangled-names -file-compilation-dir /workspaces/postgres-nio -Xcc -fmodule-map-file=/workspaces/postgres-nio/.build/checkouts/swift-atomics/Sources/_AtomicsShims/include/module.modulemap -Xcc -I -Xcc /workspaces/postgres-nio/.build/checkouts/swift-atomics/Sources/_AtomicsShims/include -Xcc -fPIC -Xcc -g -Xcc -fno-omit-frame-pointer -module-name _ConnectionPoolModule -plugin-path /usr/lib/swift/host/plugins -plugin-path /usr/local/lib/swift/host/plugins -parse-as-library -o /workspaces/postgres-nio/.build/x86_64-unknown-linux-gnu/debug/_ConnectionPoolModule.build/PoolStateMachine.swift.o -index-store-path /workspaces/postgres-nio/.build/x86_64-unknown-linux-gnu/debug/index/store -index-system-modules
1.      Swift version 6.0-dev (LLVM b83e97edcf97a27, Swift 6a1eed7489cc98f)
2.      Compiling with effective version 5.10
3.      While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "/workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine.swift")
4.      While silgen emitFunction SIL function "@$s21_ConnectionPoolModule0B12StateMachineV0A6ActionOAASQR3_rlE2eeoiySbAEyxq_q0_q1_q2_q3__G_AGtFZ".
 for '==(_:_:)' (at /workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine.swift:572:5)
5.      While verifying SIL function "@$s21_ConnectionPoolModule0B12StateMachineV0A6ActionOAASQR3_rlE2eeoiySbAEyxq_q0_q1_q2_q3__G_AGtFZ".
 for '==(_:_:)' (at /workspaces/postgres-nio/Sources/ConnectionPoolModule/PoolStateMachine.swift:572:5)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend 0x000055e5f8affae7
1  swift-frontend 0x000055e5f8afd83e
2  swift-frontend 0x000055e5f8b0015a
3  libc.so.6      0x00007f848c3c1520
4  libc.so.6      0x00007f848c4159fc pthread_kill + 300
5  libc.so.6      0x00007f848c3c1476 raise + 22
6  libc.so.6      0x00007f848c3a77f3 abort + 211
7  swift-frontend 0x000055e5f3072d77
8  swift-frontend 0x000055e5f309abab
9  swift-frontend 0x000055e5f307970c
10 swift-frontend 0x000055e5f3077b60
11 swift-frontend 0x000055e5f3075f99
12 swift-frontend 0x000055e5f306e286
13 swift-frontend 0x000055e5f24ef01e
14 swift-frontend 0x000055e5f24ee63c
15 swift-frontend 0x000055e5f24ef882
16 swift-frontend 0x000055e5f24ec9d3
17 swift-frontend 0x000055e5f26078ed
18 swift-frontend 0x000055e5f2602a48
19 swift-frontend 0x000055e5f260299d
20 swift-frontend 0x000055e5f24f2b6f
21 swift-frontend 0x000055e5f25e92a4
22 swift-frontend 0x000055e5f24f84a2
23 swift-frontend 0x000055e5f24f391c
24 swift-frontend 0x000055e5f1d6acbb
25 swift-frontend 0x000055e5f1d819c5
26 swift-frontend 0x000055e5f1d6ec07
27 swift-frontend 0x000055e5f1d6cf8c
28 swift-frontend 0x000055e5f1b2a45e
29 libc.so.6      0x00007f848c3a8d90
30 libc.so.6      0x00007f848c3a8e40 __libc_start_main + 128
31 swift-frontend 0x000055e5f1b29305
```

### Expected behavior

Compiler doesn't crash

### Environment

Swift version 6.0-dev (LLVM b83e97edcf97a27, Swift 6a1eed7489cc98f)
Target: x86_64-unknown-linux-gnu

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Snowy1803**
An alloc_stack, just like a debug_value, should ignore location overrides for their variable.

Additionally, for this to work correctly, we need to use the variable's location and not the instruction's location when generating anonymous variables in IRGen (which uses getDecl).

To implement this more easily, the variable's location is always set, and when there is no override, the duplicate location information is removed in AllocStackInst::create.

Fixes #73338
rdar://127348128

### CODE REVIEW

**Comment by Snowy1803** [comment]
@swift-ci please test

**Comment by Snowy1803** [comment]
@swift-ci please test

File: test/SILOptimizer/assemblyvision_remark/cast_remarks.swift [hunk 1]
```diff
@@ -73,14 +70,11 @@ public func condCast3<NS, T>(_ ns: NS) -> T? {
...
-  //
-  // Today, we lose that x was assigned ns2. This is flow sensitive information
-  // that we might be able to recover. We still emit that a runtime cast
-  // occurred here, just don't say what the underlying value was.
   var x = ns
   x = ns2
   return x as? T  // expected-remark @:12 {{conditional runtime cast of value with type 'NS' to 'T'}}
-                  // expected-note @-9:42 {{of 'ns2'}}
+                  // expected-note @-5:42 {{of 'ns2'}}
+                  // expected-note @-4:7 {{of 'x'}}
```
Comment (adrian-prantl) [inline test/SILOptimizer/assemblyvision_remark/cast_remarks.swift:77]: Nice, so this actually fixes another bug?

**Review by adrian-prantl — APPROVED** [review]

File: test/SILOptimizer/assemblyvision_remark/cast_remarks.swift (hunk 1 above)
Comment (Snowy1803) [reply test/SILOptimizer/assemblyvision_remark/cast_remarks.swift:77]: Yes, now that getDecl() returns the correct variable declaration after a store, this pass is able to point back to the `x` value! After optimizations. the SILValue will be associated with both debug variables, so the pass prints both of them.

**Pull request merged**